File: fbscraper.py
import os
import sys
import json
import pandas as pd
import logging
import re
from datetime import datetime
from flask import Flask, request, jsonify

app = Flask(__name__)

# --- CONFIGURATION ---
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
UPLOAD_PATH = os.path.join(BASE_DIR, "UPLOAD_FOLDER")

if not os.path.exists(UPLOAD_PATH):
    os.makedirs(UPLOAD_PATH)

# --- CSV HEADERS ---
FB_POST_HEADER = [
    'post_id', 'post_url', 'author_name', 'author_id', 
    'publish_time', 'text_content', 
    'like_count', 'love_count', 'haha_count', 'wow_count', 'sad_count', 'angry_count',
    'share_count', 'comment_count', 
    'video_view_count', 'video_duration_sec', 'video_thumbnail_url', 'media_urls'
]

FB_USER_HEADER = [
    'user_id', 'user_name', 'profile_url', 'profile_pic_url', 'is_verified'
]

# --- HELPERS ---

def find_node_by_key(obj, target_key):
    """Recursively searches for a dictionary key."""
    if isinstance(obj, dict):
        if target_key in obj: return obj[target_key]
        for key, value in obj.items():
            result = find_node_by_key(value, target_key)
            if result is not None: return result
    elif isinstance(obj, list):
        for item in obj:
            result = find_node_by_key(item, target_key)
            if result is not None: return result
    return None

def find_profile_pic(story):
    """Aggressively searches for a profile picture URI."""
    # 1. Check Context Layout (Feed)
    context_layout = find_node_by_key(story, 'context_layout')
    if context_layout:
        actor_photo = find_node_by_key(context_layout, 'actor_photo')
        if actor_photo:
            pic = find_node_by_key(actor_photo, 'profile_picture')
            if pic and isinstance(pic, dict) and 'uri' in pic:
                return pic['uri']

    # 2. Check 'actors' list (Search Results)
    actors = find_node_by_key(story, 'actors')
    if isinstance(actors, list):
        for actor in actors:
            if 'profile_picture' in actor:
                return actor['profile_picture'].get('uri')

    # 3. Fallback
    pic_node = find_node_by_key(story, 'profile_picture')
    if pic_node and isinstance(pic_node, dict) and 'uri' in pic_node:
        return pic_node['uri']
    return ""

def extract_text_content(story):
    if 'message' in story and isinstance(story['message'], dict):
        return story['message'].get('text', '')
    msg_node = find_node_by_key(story, 'message')
    if isinstance(msg_node, dict):
        return msg_node.get('text', '')
    return ""

def extract_metrics(story):
    metrics = {
        'like': 0, 'love': 0, 'haha': 0, 'wow': 0, 'sad': 0, 'angry': 0,
        'share': 0, 'comment': 0, 'views': 0
    }
    
    ufi_renderer = find_node_by_key(story, 'comet_ufi_summary_and_actions_renderer')
    if not ufi_renderer:
        feedback_context = find_node_by_key(story, 'feedback_context')
        if feedback_context:
            ufi_renderer = find_node_by_key(feedback_context, 'feedback')

    if ufi_renderer and isinstance(ufi_renderer, dict):
        feedback = ufi_renderer.get('feedback') or ufi_renderer
        
        rc = feedback.get('reaction_count')
        metrics['like'] = rc.get('count', 0) if isinstance(rc, dict) else (rc or 0)
        
        sc = feedback.get('share_count')
        metrics['share'] = sc.get('count', 0) if isinstance(sc, dict) else (sc or 0)
        
        cc = find_node_by_key(feedback, 'total_count')
        metrics['comment'] = cc if cc else 0
        
        vc = feedback.get('video_view_count')
        metrics['views'] = vc if vc else 0

        reaction_map = {
            '1635855486666999': 'like', '1678524932434102': 'love', 
            '115940658764963': 'haha', '478547315650144': 'wow', 
            '908563459236466': 'sad', '444813342392137': 'angry'
        }
        top_reactions = feedback.get('top_reactions', {}).get('edges', [])
        for edge in top_reactions:
            node = edge.get('node', {})
            r_id = node.get('id')
            count = edge.get('reaction_count', 0)
            if r_id in reaction_map:
                metrics[reaction_map[r_id]] = count
    return metrics

def extract_media_deep(story):
    """
    Scans for media URLs and video stats.
    """
    data = {
        'media_urls': set(),
        'video_duration': 0,
        'video_thumbnail': ''
    }
    
    def scan(obj):
        if isinstance(obj, dict):
            # Check for MEDIA objects
            if 'media' in obj and isinstance(obj['media'], dict):
                media = obj['media']
                
                # Images
                if 'image' in media and 'uri' in media['image']:
                    data['media_urls'].add(media['image']['uri'])
                if 'photo_image' in media and 'uri' in media['photo_image']:
                    data['media_urls'].add(media['photo_image']['uri'])
                if 'large_share_image' in media and 'uri' in media['large_share_image']:
                    data['media_urls'].add(media['large_share_image']['uri'])

                # Videos (Progressive is best)
                if 'videoDeliveryResponseFragment' in media:
                    res = media['videoDeliveryResponseFragment'].get('videoDeliveryResponseResult')
                    if res and 'progressive_urls' in res:
                        for prog in res['progressive_urls']:
                            if 'progressive_url' in prog:
                                data['media_urls'].add(prog['progressive_url'])
                
                # Video Fallbacks
                if 'playable_url' in media and media['playable_url']:
                    data['media_urls'].add(media['playable_url'])
                if 'browser_native_hd_url' in media and media['browser_native_hd_url']:
                    data['media_urls'].add(media['browser_native_hd_url'])
                if 'browser_native_sd_url' in media and media['browser_native_sd_url']:
                    data['media_urls'].add(media['browser_native_sd_url'])

                # Thumbnails
                if 'preferred_thumbnail' in media and 'image' in media['preferred_thumbnail']:
                    uri = media['preferred_thumbnail']['image'].get('uri')
                    if uri:
                        data['video_thumbnail'] = uri
                        data['media_urls'].add(uri)

                # Duration
                if 'playable_duration_in_ms' in media:
                    try: data['video_duration'] = media['playable_duration_in_ms'] / 1000
                    except: pass

            # Recurse values
            for k, v in obj.items():
                scan(v)
                
        elif isinstance(obj, list):
            for item in obj:
                scan(item)

    scan(story)
    return data

# --- MAIN LOGIC ---

def extract_fb_data(har_data):
    posts = {}
    users = {}

    entries = har_data.get('log', {}).get('entries', [])
    for entry in entries:
        if "facebook.com/api/graphql/" not in entry['request']['url']:
            continue
            
        resp_text = entry.get('response', {}).get('content', {}).get('text', '')
        if not resp_text: continue

        parts = resp_text.strip().split('\n')
        for part in parts:
            try:
                data = json.loads(part)
                
                # --- A. USER EXTRACTION (Search Results) ---
                if 'serpResponse' in data.get('data', {}):
                     results = data['data']['serpResponse'].get('results', {}).get('edges', [])
                     for res in results:
                         if 'rendering_strategy' in res:
                             strategies = res['rendering_strategy'].get('result_rendering_strategies', [])
                             for strat in strategies:
                                 vm = strat.get('view_model', {})
                                 prof = vm.get('profile', {})
                                 if prof:
                                     u_id = prof.get('id')
                                     if u_id and u_id not in users:
                                         users[u_id] = {
                                             'user_id': u_id,
                                             'user_name': prof.get('name'),
                                             'profile_url': prof.get('url'),
                                             'profile_pic_url': find_profile_pic(prof),
                                             'is_verified': prof.get('is_verified', False)
                                         }

                # --- B. POST EXTRACTION ---
                edges = []
                if 'serpResponse' in data.get('data', {}):
                    edges = data['data']['serpResponse'].get('results', {}).get('edges', [])
                elif 'node' in data.get('data', {}):
                    edges = [{'node': data['data']['node']}]
                elif 'viewer' in data.get('data', {}):
                    edges = data['data']['viewer'].get('news_feed', {}).get('edges', [])
                
                if not edges:
                    found_edges = find_node_by_key(data, 'edges')
                    if isinstance(found_edges, list): edges = found_edges

                if not edges: continue

                for edge in edges:
                    if not isinstance(edge, dict): continue
                    
                    story = find_node_by_key(edge, 'story')
                    if not story or not isinstance(story, dict): continue
                    
                    p_id = story.get('post_id') or story.get('id')
                    if not p_id or len(str(p_id)) < 5: continue 
                    if p_id in posts: continue

                    # 1. Author
                    actor = {}
                    actors_list = story.get('actors') or find_node_by_key(story, 'actors')
                    if actors_list and isinstance(actors_list, list) and len(actors_list) > 0:
                        actor = actors_list[0]
                    
                    u_id = actor.get('id')
                    u_name = actor.get('name')
                    u_url = actor.get('url')
                    u_pic = find_profile_pic(story)
                    
                    # 2. Verification
                    u_verified = False
                    title_ranges = find_node_by_key(story, 'ranges')
                    if title_ranges and isinstance(title_ranges, list):
                        for rng in title_ranges:
                            entity = rng.get('entity', {})
                            if entity.get('id') == u_id:
                                u_verified = entity.get('is_verified', False)
                    if not u_verified and actor.get('is_verified'):
                        u_verified = True

                    # 3. Metadata
                    p_url = story.get('permalink_url') or story.get('wwwURL') or story.get('url')
                    if not p_url: p_url = f"https://www.facebook.com/{p_id}"

                    publish_time = ""
                    creation_time = story.get('creation_time') or find_node_by_key(story, 'creation_time')
                    if creation_time:
                        try: publish_time = datetime.fromtimestamp(int(creation_time)).strftime('%Y-%m-%d %H:%M:%S')
                        except: pass

                    text_content = extract_text_content(story)
                    metrics = extract_metrics(story)
                    
                    # 4. MEDIA & LINKS (Deep Scan)
                    media_data = extract_media_deep(story)
                    clean_media = "; ".join(list(media_data['media_urls']))

                    posts[p_id] = {
                        'post_id': p_id,
                        'post_url': p_url,
                        'author_name': u_name,
                        'author_id': u_id,
                        'publish_time': publish_time,
                        'text_content': text_content.replace('\n', ' ').strip(),
                        
                        'like_count': metrics['like'],
                        'love_count': metrics['love'],
                        'haha_count': metrics['haha'],
                        'wow_count': metrics['wow'],
                        'sad_count': metrics['sad'],
                        'angry_count': metrics['angry'],
                        'share_count': metrics['share'],
                        'comment_count': metrics['comment'],
                        
                        'video_view_count': metrics['views'],
                        'video_duration_sec': media_data['video_duration'],
                        'video_thumbnail_url': media_data['video_thumbnail'],
                        'media_urls': clean_media
                    }

                    if u_id and u_id not in users:
                        users[u_id] = {
                            'user_id': u_id,
                            'user_name': u_name,
                            'profile_url': u_url,
                            'profile_pic_url': u_pic,
                            'is_verified': u_verified
                        }
                    elif u_id and u_verified and not users[u_id]['is_verified']:
                        users[u_id]['is_verified'] = True
                             
            except Exception: continue
            
    return list(posts.values()), list(users.values())

def run_processing(filename):
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            har_data = json.load(f)

        posts, users = extract_fb_data(har_data)
        
        base_name = os.path.splitext(os.path.basename(filename))[0]
        dt_postfix = datetime.now().strftime('%Y%m%d_%H%M%S')
        
        p_csv = f"{base_name}_fbposts_{dt_postfix}.csv"
        u_csv = f"{base_name}_fbusers_{dt_postfix}.csv"

        if posts:
            pd.DataFrame(posts).reindex(columns=FB_POST_HEADER).to_csv(os.path.join(UPLOAD_PATH, p_csv), index=False)
        else:
            pd.DataFrame(columns=FB_POST_HEADER).to_csv(os.path.join(UPLOAD_PATH, p_csv), index=False)
            
        if users:
            pd.DataFrame(users).reindex(columns=FB_USER_HEADER).to_csv(os.path.join(UPLOAD_PATH, u_csv), index=False)
        else:
            pd.DataFrame(columns=FB_USER_HEADER).to_csv(os.path.join(UPLOAD_PATH, u_csv), index=False)

        return {
            "status": "success",
            "posts_count": len(posts),
            "users_count": len(users),
            "files": [p_csv, u_csv]
        }
    except Exception as e:
        return {"status": "error", "message": str(e)}

@app.route("/api/process", methods=["POST"])
def process_api():
    if 'file' not in request.files: return jsonify({"error": "No file"}), 400
    file = request.files['file']
    if file.filename == '': return jsonify({"error": "No file"}), 400
    filepath = os.path.join(UPLOAD_PATH, file.filename)
    file.save(filepath)
    return jsonify(run_processing(filepath))

if __name__ == "__main__":
    if len(sys.argv) > 1:
        print(json.dumps(run_processing(sys.argv[1])))
    else:
        app.run(host='0.0.0.0', port=5000)

File: fbscraper.php
<?php
/**
 * fbscraper.php - Core Logic
 * Converts HAR data to CSV using native PHP.
 */

function findAllNodes($obj, $targetKeys, $maxDepth = 50)
{
    if (!is_array($obj) && !is_object($obj))
        return [];

    $results = [];
    $stack = [['o' => $obj, 'd' => 0]];

    while (count($stack) > 0) {
        $item = array_pop($stack);
        $o = $item['o'];
        $d = $item['d'];

        if (!is_array($o) && !is_object($o))
            continue;
        if ($d > $maxDepth)
            continue;

        foreach ($o as $key => $val) {
            if (in_array($key, $targetKeys)) {
                $results[] = $val;
            }
            if (is_array($val) || is_object($val)) {
                $stack[] = ['o' => $val, 'd' => $d + 1];
            }
        }
    }
    return $results;
}

function extractFeedback($node)
{
    // Try to find the most relevant feedback object
    // It's often nested under 'feedback', 'comet_ufi_summary_and_actions_renderer', etc.
    // Increase depth significantly to catch nested feedback structures
    $feedbacks = findAllNodes($node, ['feedback'], 40);

    $stats = [
        'likes' => 0,
        'love' => 0,
        'haha' => 0,
        'wow' => 0,
        'sad' => 0,
        'angry' => 0,
        'care' => 0,
        'comments' => 0,
        'shares' => 0
    ];

    foreach ($feedbacks as $f) {
        if (isset($f['reaction_count']['count'])) {
            $stats['likes'] = max($stats['likes'], (int)$f['reaction_count']['count']);
        }
        elseif (isset($f['i18n_reaction_count'])) {
            // Handle "1.2K" etc strictly if needed, but mostly it's raw numbers in JSON
            $stats['likes'] = max($stats['likes'], (int)filter_var($f['i18n_reaction_count'], FILTER_SANITIZE_NUMBER_INT));
        }

        // Detailed Reactions (Top Reactions)
        if (isset($f['top_reactions']['edges']) && is_array($f['top_reactions']['edges'])) {
            foreach ($f['top_reactions']['edges'] as $edge) {
                // Determine name based on localized_name (fallback to safe defaults if English)
                // Note: Localized names depend on the user's locale in HAR. 
                // We assume standard English names or match loosely.
                $reactionName = strtolower($edge['node']['localized_name'] ?? '');
                $count = (int)($edge['reaction_count'] ?? 0);

                if (isset($stats[$reactionName])) {
                    $stats[$reactionName] = max($stats[$reactionName], $count);
                }
            }
        }

        // Comment Count Detection
        if (isset($f['total_comment_count'])) {
            $stats['comments'] = max($stats['comments'], (int)$f['total_comment_count']);
        }
        elseif (isset($f['comment_count']['total_count'])) {
            $stats['comments'] = max($stats['comments'], (int)$f['comment_count']['total_count']);
        }
        else {
            // Check for comet style nesting: comment_rendering_instance -> comments -> total_count
            $deepComments = findAllNodes($f, ['total_count'], 5);
            foreach ($deepComments as $dc) {
                if (is_numeric($dc))
                    $stats['comments'] = max($stats['comments'], (int)$dc);
            }
        }

        if (isset($f['share_count']['count'])) {
            $stats['shares'] = max($stats['shares'], (int)$f['share_count']['count']);
        }
        elseif (isset($f['share_count_reduced'])) {
            $stats['shares'] = max($stats['shares'], (int)filter_var($f['share_count_reduced'], FILTER_SANITIZE_NUMBER_INT));
        }
    }
    return $stats;
}

function findDate($node)
{
    // Search deep for time keys (up to depth 15, to catch deeply nested creation_time)
    $candidates = findAllNodes($node, ['creation_time', 'publish_time'], 15);

    $best = 0;
    foreach ($candidates as $time) {
        if (is_numeric($time) && $time > 0) {
            // Filter out future dates or impossibly old dates
            // Assuming HAR is recent (2000-2030)
            // But sometimes timestamps are milliseconds.
            // If > 20000000000 (year ~2600), treat as ms.
            if ($time > 20000000000)
                $time = $time / 1000;

            // We want the most plausible timestamp.
            // Usually any valid timestamp is good.
            if ($best == 0)
                $best = $time;
            else {
                // Heuristic: prefer earlier time? Or later?
                // Often 'creation_time' is earlier than 'publish_time' for edits?
                // Let's stick to first valid found for simplicity, or max?
                // Let's prefer the one closest to now? 
                // Or simply max to get the latest update time?
                // Actually, creation_time is usually what we want.
                // Let's take the minimum non-zero value to get the original creation time?
                if ($time < $best)
                    $best = $time;
            }
        }
    }
    return $best;
}

function extractUserProfile($actor)
{
    $profile = [
        'id' => $actor['id'] ?? '',
        'name' => $actor['name'] ?? ($actor['text'] ?? ($actor['label'] ?? ($actor['short_name'] ?? 'Unknown'))),
        'url' => $actor['url'] ?? ($actor['uri'] ?? ("https://facebook.com/" . ($actor['id'] ?? ''))),
        'is_verified' => (isset($actor['is_verified']) && $actor['is_verified']) || 
                         (isset($actor['verification_status']) && $actor['verification_status'] === 'VERIFIED') ||
                         (isset($actor['__typename']) && strpos($actor['__typename'], 'VerifiedBadge') !== false)
                         ? 'Yes' : 'No',
        'pic' => ''
    ];

    // Aggressive search for picture URI - prioritize specific keys first
    $specificKeys = ['profile_picture', 'profile_photo', 'profile_pic', 'displayPicture', 'avatar', 'profileImage'];
    $genericKeys = ['image', 'picture'];

    // First pass: Specific Keys
    $nodes = findAllNodes($actor, $specificKeys, 15);
    foreach ($nodes as $n) {
        if (is_array($n)) {
            $uri = $n['uri'] ?? ($n['url'] ?? ($n['src'] ?? ''));
            if ($uri && strpos($uri, 'http') === 0) {
                $profile['pic'] = $uri;
                break;
            }
        }
        elseif (is_string($n) && strpos($n, 'http') === 0) {
            $profile['pic'] = $n;
            break;
        }
    }

    // Second pass: Generic Keys (only if no pic found yet)
    if (empty($profile['pic'])) {
        $nodes = findAllNodes($actor, $genericKeys, 10);
        foreach ($nodes as $n) {
            if (is_array($n)) {
                $uri = $n['uri'] ?? ($n['url'] ?? ($n['src'] ?? ''));
                if ($uri && strpos($uri, 'http') === 0) {
                    $profile['pic'] = $uri;
                    break;
                }
            }
            elseif (is_string($n) && strpos($n, 'http') === 0) {
                $profile['pic'] = $n;
                break;
            }
        }
    }

    return $profile;
}

function extractMedia($node)
{
    $media = [];
    $seen = [];

    // Search DEEPLY for specific image keys directly, bypassing attachment wrappers
    // which seemed unreliable in some HAR structures.
    // Depth 50 to catch everything.
    $images = findAllNodes($node, ['image', 'photo_image', 'original_image', 'large_share_image'], 50);

    foreach ($images as $img) {
        $uri = $img['uri'] ?? ($img['url'] ?? ($img['src'] ?? null));

        // Filter out tiny tracking pixels or icons if needed
        // (But usually FB image URLs are distinct)
        if ($uri && !isset($seen[$uri]) && strpos($uri, 'http') === 0) {
            $media[] = $uri;
            $seen[$uri] = true;
        }
    }

    // Direct video URLs
    // playable_url_quality_hd is often a direct key in the node
    $videos = findAllNodes($node, ['playable_url_quality_hd', 'playable_url'], 50);
    foreach ($videos as $v) {
        if (is_string($v) && strpos($v, 'http') === 0 && !isset($seen[$v])) {
            $media[] = $v;
            $seen[$v] = true;
        }
    }

    return implode("; ", $media);
}

function processHarFile($harPath, $uploadDir)
{
    $harName = pathinfo($harPath, PATHINFO_FILENAME);
    $timestamp = date("YmdHis");

    $jsonData = json_decode(file_get_contents($harPath), true);
    if (!$jsonData)
        return ["status" => "error", "message" => "Invalid JSON in HAR file"];

    $posts = [];
    $users = [];

    $entries = $jsonData['log']['entries'] ?? [];

    foreach ($entries as $entry) {
        $text = $entry['response']['content']['text'] ?? '';
        if (empty($text))
            continue;

        $dataObjects = [];

        $cleanText = $text;
        if (substr($text, 0, 9) === "for (;;);") {
            $cleanText = substr($text, 9);
        }

        $firstTry = json_decode($cleanText, true);
        if ($firstTry) {
            $dataObjects[] = $firstTry;
        }
        else {
            $lines = explode("\n", $text);
            foreach ($lines as $line) {
                if (trim($line) === '')
                    continue;
                $d = json_decode($line, true);
                if ($d)
                    $dataObjects[] = $d;
            }
        }

        foreach ($dataObjects as $data) {
            // 1. Global User Discovery (Search for user-like objects anywhere in the data)
            $userKeys = ['user', 'actor', 'actors', 'author', 'participant', 'owner', 'entity', 'owning_profile', 'creator', 'messaging_sender'];
            $potentialUsers = findAllNodes($data, $userKeys, 20);

            foreach ($potentialUsers as $u) {
                $items = is_array($u) && isset($u[0]) ? $u : [$u];
                foreach ($items as $actor) {
                    if (!is_array($actor) || empty($actor['id']) || !is_numeric($actor['id']))
                        continue;

                    $profile = extractUserProfile($actor);
                    $uid = $actor['id'];

                    // Skip Anonymous participants or things that look like Hashtags/Groups
                    $nameLower = strtolower($profile['name']);
                    if (strpos($nameLower, 'anonymous participant') !== false ||
                    strpos($profile['url'], '/hashtag/') !== false ||
                    strpos($profile['url'], '/groups/') !== false) {
                        continue;
                    }

                    // For broad discovery, if we have NO name and NO pic, it's probably junk
                    // UNLESS they are verified!
                    if (($profile['name'] === 'Unknown' || empty($profile['name'])) && empty($profile['pic']) && $profile['is_verified'] !== 'Yes') {
                        continue;
                    }

                    $shouldUpdate = false;
                    if (!isset($users[$uid])) {
                        $shouldUpdate = true;
                    }
                    elseif (empty($users[$uid]['profile_pic']) && !empty($profile['pic'])) {
                        $shouldUpdate = true;
                    }
                    elseif (($users[$uid]['is_verified'] ?? 'No') === 'No' && $profile['is_verified'] === 'Yes') {
                        $shouldUpdate = true;
                    }

                    if ($shouldUpdate) {
                        $existing = $users[$uid] ?? [];
                        $users[$uid] = [
                            'user_id' => $profile['id'],
                            'user_name' => !empty($profile['name']) && $profile['name'] !== 'Unknown' ? $profile['name'] : ($existing['user_name'] ?? 'Unknown'),
                            'profile_url' => $profile['url'],
                            'is_verified' => $profile['is_verified'] === 'Yes' ? 'Yes' : ($existing['is_verified'] ?? 'No'),
                            'profile_pic' => !empty($profile['pic']) ? $profile['pic'] : ($existing['profile_pic'] ?? '')
                        ];
                    }
                }
            }

            // 2. Post Extraction
            $nodes = array_merge(
                findAllNodes($data, ['node']),
                findAllNodes($data, ['story'])
            );

            foreach ($nodes as $n) {
                // Extract Post Data (Strictly BigInt IDs only)
                $id = $n['post_id'] ?? ($n['id'] ?? null);
                if (!$id || strlen((string)$id) < 5 || !is_numeric($id))
                    continue;

                // Ensure it's a valid post content-like node
                if (!isset($n['actors']) && !isset($n['message']) && !isset($n['comet_sections']) && !isset($n['feedback']))
                    continue;

                // Extract metrics and media
                $metrics = extractFeedback($n);
                $mediaUrls = extractMedia($n);
                $timestamp_post = findDate($n);

                // Deep extract text body
                $textBody = $n['message']['text'] ?? "";
                if (empty($textBody)) {
                    $allText = findAllNodes($n, ['message'], 20);
                    foreach ($allText as $mt) {
                        if (isset($mt['text']) && !empty($mt['text'])) {
                            $textBody = $mt['text'];
                            break;
                        }
                    }
                }

                $author = findAllNodes($n, ['actors'])[0][0] ?? [];
                
                // Enhanced Verification Check for Comet
                $isAuthorVerified = (isset($author['is_verified']) && $author['is_verified']) || (isset($author['verification_status']) && $author['verification_status'] === 'VERIFIED');
                if (!$isAuthorVerified) {
                    $typenames = findAllNodes($n, ['__typename'], 15);
                    foreach ($typenames as $tn) {
                        if (is_string($tn) && strpos($tn, 'VerifiedBadge') !== false) {
                            $isAuthorVerified = true;
                            break;
                        }
                    }
                }

                // Update user global list if we found a verified status here
                if ($isAuthorVerified && !empty($author['id'])) {
                    $auid = $author['id'];
                    if (isset($users[$auid])) {
                        $users[$auid]['is_verified'] = 'Yes';
                    }
                }

                if (!isset($posts[$id])) {
                    $posts[$id] = [
                        'post_id' => $id,
                        'post_url' => $n['url'] ?? "https://facebook.com/$id",
                        'author_name' => $author['name'] ?? "Unknown",
                        'author_id' => (isset($author['id']) && is_numeric($author['id'])) ? $author['id'] : "",
                        'text_content' => str_replace(["\r", "\n"], " ", $textBody),
                        'likes' => $metrics['likes'],
                        'love' => $metrics['love'],
                        'haha' => $metrics['haha'],
                        'wow' => $metrics['wow'],
                        'sad' => $metrics['sad'],
                        'angry' => $metrics['angry'],
                        'care' => $metrics['care'],
                        'comments' => $metrics['comments'],
                        'shares' => $metrics['shares'],
                        'media_urls' => $mediaUrls,
                        'publish_time' => $timestamp_post ? date('Y-m-d H:i:s', $timestamp_post) : ''
                    ];
                }
                else {
                    // Merge/Update existing post with better data
                    if (empty($posts[$id]['text_content']) && !empty($textBody)) {
                        $posts[$id]['text_content'] = str_replace(["\r", "\n"], " ", $textBody);
                    }
                    if (empty($posts[$id]['author_id']) && !empty($author['id'])) {
                        $posts[$id]['author_id'] = $author['id'];
                        $posts[$id]['author_name'] = $author['name'] ?? "Unknown";
                    }
                    if (empty($posts[$id]['media_urls']) && !empty($mediaUrls)) {
                        $posts[$id]['media_urls'] = $mediaUrls;
                    }
                    if (empty($posts[$id]['publish_time']) && $timestamp_post) {
                        $posts[$id]['publish_time'] = date('Y-m-d H:i:s', $timestamp_post);
                    }

                    // Update metrics to max found
                    foreach ($metrics as $mKey => $mVal) {
                        if (isset($posts[$id][$mKey]) && $mVal > $posts[$id][$mKey]) {
                            $posts[$id][$mKey] = $mVal;
                        }
                    }
                }
            }
        }
    }

    $postsFile = "{$harName}_fbposts_{$timestamp}.csv";
    $usersFile = "{$harName}_fbusers_{$timestamp}.csv";

    $fp = fopen($uploadDir . $postsFile, 'w');
    fputcsv($fp, ['post_id', 'post_url', 'author_name', 'author_id', 'text_content', 'likes', 'love', 'haha', 'wow', 'sad', 'angry', 'care', 'comments', 'shares', 'media_urls', 'publish_time'], ",", "\"", "\\");
    foreach ($posts as $post)
        fputcsv($fp, $post, ",", "\"", "\\");
    fclose($fp);

    $fu = fopen($uploadDir . $usersFile, 'w');
    fputcsv($fu, ['user_id', 'user_name', 'profile_url', 'is_verified', 'profile_pic'], ",", "\"", "\\");
    foreach ($users as $user)
        fputcsv($fu, $user, ",", "\"", "\\");
    fclose($fu);

    return ["status" => "success", "files" => [$postsFile, $usersFile]];
}